app/javascript/components/widget-controller.jsx

```diff
@@ -44,223 +44,205 @@ const widgetElements = {
   Traffic,
 };
 
-export class WidgetController extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      currentWidgetId: 0,
-      transitionTime: null,
+const prefetchDelay = 1500;
+
+export const WidgetController = ({ client, cityName }) => {
+  const lastPrefetchTime = useRef(0);
+  const timerId = useRef(null);
+  const [state, setState] = useState({
+    currentWidgetId: 0,
+    transitionTime: null,
+    pausedTime: null,
+  });
+
+  useEffect(() => {
+    return () => {
+      if (timerId.current) {
+        clearInterval(timerId.current);
+      }
     };
-  }
-
-  componentDidMount() {
-    const prefetchDelay = 1500;
+  }, []);
+
+  const fetchPolicy =
+    state.currentWidgetId === 0 ? 'network-only' : 'cache-first';
+
+  return (
+    <Query
+      query={getWidgets}
+      variables={{ id: state.currentWidgetId, cityName }}
+      fetchPolicy={fetchPolicy}
+      onCompleted={response => {
+        const { location } = response;
+        const { widgets } = location;
+        const { byIdOrFirst: current } = widgets;
+
+        setState(({ pausedTime }) => ({
+          currentWidgetId: current.id,
+          pausedTime,
+          transitionTime: Date.now() + current.durationSeconds * 1000,
+        }));
+      }}
+    >
+      {({ loading, error, data }) => {
+        if (loading) {
+          return (
+            <Wrapper>
+              <FullPanel>
+                <LoadingMessage />
+              </FullPanel>
+              <SidePanel
+                widgets={[]}
+                selectedWidgetId={0}
+                showWeather={false}
+                totalTime={0}
+                isPaused={false}
+                cityName={cityName}
+              />
+            </Wrapper>
+          );
+        }
 
-    this.lastPrefetchTime = 0;
-    this.timerId = setInterval(() => {
-      const { transitionTime, pausedTime } = this.state;
+        if (error) {
+          // eslint-disable-next-line
+          console.error(error);
+          return (
+            <Wrapper>
+              <FullPanel>
+                <DisconnectedMessage />
+              </FullPanel>
+              <SidePanel
+                widgets={[]}
+                selectedWidgetId={0}
+                showWeather={false}
+                totalTime={0}
+                isPaused={false}
+                cityName={cityName}
+              />
+            </Wrapper>
+          );
+        }
 
-      if (!transitionTime || pausedTime) {
-        return;
-      }
+        const { location } = data;
+        const { widgets } = location;
+        const { enabled: enabledWidgets, byIdOrFirst: current } = widgets;
```

Comment:
unless you need to keep these intermediate variables around for some reason you can destructure all at once like this:

```js
const { 
  location: { 
    widgets: {
      enabled: enabledWidgets, 
      byIdOrFirst: current } } } = data;
```

Reply:
I'll address that in the refactor PR